File: cache/cache.go
package cache

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/milicacurcic/bcpartizan-ics/types"
)

const cacheFile = "data/cache.json"

type CacheData struct {
	Timestamp int64        `json:"timestamp"`
	Games     []types.Game `json:"games"`
}

func Load(cacheTTL time.Duration) ([]types.Game, error) {
	data, err := os.ReadFile(cacheFile)
	if err != nil {
		return nil, err
	}

	var cache CacheData
	if err := json.Unmarshal(data, &cache); err != nil {
		return nil, err
	}

	now := time.Now().Unix()
	if cache.Timestamp > 0 && (now-cache.Timestamp) < int64(cacheTTL.Seconds()) {
		return cache.Games, nil
	}

	return nil, nil
}

func Save(games []types.Game) error {
	cache := CacheData{
		Timestamp: time.Now().Unix(),
		Games:     games,
	}

	data, err := json.MarshalIndent(cache, "", "  ")
	if err != nil {
		return err
	}

	dir := filepath.Dir(cacheFile)
	if err := os.MkdirAll(dir, 0755); err != nil {
		log.Printf("Error creating cache directory: %v", err)
		return err
	}

	if err := os.WriteFile(cacheFile, data, 0644); err != nil {
		log.Printf("Error saving cache: %v", err)
		return err
	}

	return nil
}

File: src/cache.js
const fs = require('fs').promises;
const path = require('path');
const config = require('./config');

const CACHE_FILE = path.join(__dirname, '../data/cache.json');

async function loadCache() {
  try {
    const data = await fs.readFile(CACHE_FILE, 'utf8');
    const cache = JSON.parse(data);
    
    const now = Date.now();
    if (cache.timestamp && (now - cache.timestamp) < config.cacheTTL) {
      return cache.games;
    }
    
    return null;
  } catch (error) {
    return null;
  }
}

async function saveCache(games) {
  const cache = {
    timestamp: Date.now(),
    games
  };
  
  try {
    await fs.mkdir(path.dirname(CACHE_FILE), { recursive: true });
    await fs.writeFile(CACHE_FILE, JSON.stringify(cache, null, 2));
  } catch (error) {
    console.error('Error saving cache:', error.message);
  }
}

module.exports = {
  loadCache,
  saveCache
};
